core/src/main/scala/org/bykn/bosatsu/SourceConverter.scala

```diff
@@ -58,48 +58,59 @@ final class SourceConverter(
     decl match {
       case Apply(fn, args, _) =>
-        Expr.buildApp(apply(fn), args.toList.map(apply(_)), decl)
+        (apply(fn), args.toList.traverse(apply(_)))
+          .mapN { Expr.buildApp(_, _, decl) }
       case ao@ApplyOp(left, op, right) =>
         val opVar: Expr[Declaration] = Expr.Var(None, op, ao.opVar)
-        Expr.buildApp(opVar, apply(left) :: apply(right) :: Nil, decl)
+        (apply(left), apply(right)).mapN { (l, r) =>
+          Expr.buildApp(opVar, l :: r :: Nil, decl)
+        }
       case Binding(BindingStatement(pat, value, prest@Padding(_, rest))) =>
         val erest = apply(rest)
 
-        def solvePat(pat: Pattern.Parsed, rhs: Expr[Declaration]): Expr[Declaration] =
+        def solvePat(pat: Pattern.Parsed, rrhs: Result[Expr[Declaration]]): Result[Expr[Declaration]] =
           pat match {
             case Pattern.Var(arg) =>
-              Expr.Let(arg, rhs, erest, RecursionKind.NonRecursive, decl)
+              (erest, rrhs).mapN { (e, rhs) =>
+                Expr.Let(arg, rhs, e, RecursionKind.NonRecursive, decl)
+              }
             case Pattern.Annotation(pat, tpe) =>
               val realTpe = toType(tpe)
               // move the annotation to the right
-              val newRhs = Expr.Annotation(rhs, realTpe, decl)
+              val newRhs = rrhs.map(Expr.Annotation(_, realTpe, decl))
               solvePat(pat, newRhs)
             case Pattern.Named(nm, p) =>
                // this is the same as creating a let nm = value first
-               val inner = solvePat(p, rhs)
-               Expr.Let(nm, rhs, inner, RecursionKind.NonRecursive, decl)
+              (solvePat(p, rrhs), rrhs).mapN { (inner, rhs) =>
+                Expr.Let(nm, rhs, inner, RecursionKind.NonRecursive, decl)
+              }
             case pat =>
               // TODO: we need the region on the pattern...
-              val newPattern = unTuplePattern(pat, decl.region)
-              val expBranches = NonEmptyList.of((newPattern, erest))
-              Expr.Match(rhs, expBranches, decl)
+              (unTuplePattern(pat, decl.region), erest, rrhs).mapN { (newPattern, e, rhs) =>
+                val expBranches = NonEmptyList.of((newPattern, e))
+                Expr.Match(rhs, expBranches, decl)
+              }
           }
 
         solvePat(pat, apply(value))
       case Comment(CommentStatement(_, Padding(_, decl))) =>
-        apply(decl).map(_ => decl)
+        apply(decl).map(_.as(decl))
       case DefFn(defstmt@DefStatement(_, _, _, _)) =>
         val inExpr = defstmt.result match {
           case (_, Padding(_, in)) => apply(in)
         }
-        val lambda = defstmt.toLambdaExpr({ res => apply(res._1.get) }, decl)(
-          unTuplePattern(_, decl.region), toType(_))
+        // TODO
+        val lambda = defstmt.toLambdaExprF({ res => apply(res._1.get) }, success(decl))(
+          unTuplePattern(_, decl.region), { t => success(toType(t)) })
```

Comment:
we should lift `toType` into `Result` by checking the visible types, and either return `Ior.Right` or `Ior.Both` to signal failure but also check to see if we can find any other errors.